transport/transport.go

```diff
@@ -0,0 +1,113 @@
+package transport
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Transport interface {
+	// NamespacedName returns the namespaced name to identify this transport Transport
+	NamespacedName() types.NamespacedName
+	// ListenPort returns a port on which the transport server listens for incomming connections
+	ListenPort() int32
+	// ConnectPort returns the port to connect to transfer server
+	// Using this the server acts as a client to the transfer relaying all the data
+	// sent from the transport client
+	ConnectPort() int32
+	// Containers returns a list of containers transfers can add to their server Pods
+	Containers() []corev1.Container
+	// Volumes returns a list of volumes transfers have add to their server Pods for getting the configurations
+	// mounted for the transport containers to work
+	Volumes() []corev1.Volume
+	// Type
+	Type() Type
+	// Credentials returns the namespaced name of the secret holding credentials for talking to the server
+	Credentials() types.NamespacedName
+	// Hostname returns the string to which the transfer will connect to
+	// in case of a null transport, it will simple relay the endpoint hostname
+	// in case of a valid transport, it will have a custom hostname where transfers will have to connect to.
+	Hostname() string
```

Comment:
I'm having trouble understanding `ListenPort`, `ConnectPort`, and `Hostname`. Should I be reading this as:
- For server: We expect the incoming data on hostname:listenport and it unwraps and sends to localhost:connectport
- For client: We expect incoming data on localhost:listenport and it wraps and sends to hostname:connectport

Reply:
Hostname() is for the transfer to consume. In case of client, the transfer client will connect using hostname:listenport and in case of server, the transfer server will allow for connection from hostname.

Since the stunnel implementation we are choosing is using sidecar container, we will always return "localhost" as the hostname, but you could imagine a different implementation where 1 stunnel pod and service is multiplexing to a different rsyncd server, in which case the hostname would be the service dns.

tldr:
1. Hostname() is just for communication between transfer and transport
2. ListenPort and ConnectPort will differ based on server and client you have pointed out.
3. The public dnsname/ip address to which client transport connects to the server will come in as a parameter to the client.New